README.md
# Springpunk Migration Tools (SPMT) for Unity

Editor tools to facilitate some steps in migrating/updating to new systems

![Screenshot](images/screenshot.png)

# Default functionalities

- Project scene selection
- Scene folder and subfolder creation
- Opening scenes, marking them as dirty and saving them
- Custom tool creation

# How to install

- Open the Unity Package Manager (`Window > Package Manager`)
- Click the "plus" icon in the top left
- Select "Add package from git URL"
- Paste in the git URL for this repo: `https://github.com/springpunk/unity-migration-tools.git`

# Usag

## Opening the window

- `Window > Springpunk > Migration Tools`

## Getting the loaded scenes from the SPMT Window

```cs
using Springpunk.MigrationTools.Editor;

public class Example {

    public void Foo(){
        SceneLoader sceneLoader = MigrationToolsWindow.Instance.SceneLoader;
        string[] selectedScenePaths = sceneLoader.SelectedScenePaths
    }
}
```

## Creating a custom tool

Check the [example](Example/ExampleMigrationTool.cs) to see how to implement custom tools

```cs
using UnityEditor;
using Springpunk.MigrationTools.Editor;

[InitializeOnLoad]
public class MyCustomTool : MigrationTool
{
    // Automatically register the tool for use on editor load
    static MyCustomTool()
    {
        MyCustomTool example = new MyCustomTool();
        MigrationToolsWindow.RegisterCustomTool(example);
    }

    public MyCustomTool() : base("My Custom Tool") { }

    protected override void OnGUI()
    {
        // Editor GUI code goes here...
    }
}

```


CHANGELOG.md
# Changelog

## 1.0.2

- Made a scroll view for the scene loader


Editor/SceneLoader.cs
using System.Linq;
using System.Collections.Generic;
using UnityEngine;
using UnityEditor;

namespace Springpunk.MigrationTools.Editor
{
    internal class SceneLoadItem {
        public string scenePath;
        public bool select;

        public SceneLoadItem(string scenePath, bool select = true) {
            this.scenePath = scenePath;
            this.select = select;
        }
    }

    public class SceneLoader : MigrationTool
    {
        private List<SceneLoadItem> sceneLoadItems = new List<SceneLoadItem>();
        public string[] SelectedScenePaths {
            get {
                
                string[] paths = sceneLoadItems
                    .Where((item) => item.select).ToList()
                    .ConvertAll((item) => item.scenePath).ToArray();
                if (paths.Length < 1) return new string[0];
                return paths;
            }
        }

        private bool showScenes = false;
        private Vector2 scrollProgress = Vector2.zero;

        public SceneLoader() : base("Scene Loader") { }

        protected override void OnGUI()
        {
            GUILayout.Label("Scene loading:");
            if (GUILayout.Button("Get All Scenes"))
            {
                LoadAllScenes();
            }

            if (sceneLoadItems.Count < 1) {
                GUILayout.Label("No scenes loaded in the tool...");
                return;
            }

            int selectedCount = CountSelectedScenes();
            string sceneCountLabel = $"Selected scenes: {selectedCount}/{sceneLoadItems.Count}";
            showScenes = EditorGUILayout.BeginFoldoutHeaderGroup(showScenes, sceneCountLabel);
            if (showScenes)
            {
                GUILayout.BeginHorizontal();
                if (GUILayout.Button("Select all"))
                    sceneLoadItems.ForEach((item) => item.select = true);
                if(GUILayout.Button("Select none"))
                    sceneLoadItems.ForEach((item) => item.select = false);
                GUILayout.EndHorizontal();
                scrollProgress = GUILayout.BeginScrollView(scrollProgress, GUILayout.MaxHeight(400));
                for (int i = 0; i < sceneLoadItems.Count; i++)
                {
                    GUILayout.BeginHorizontal("box");
                    sceneLoadItems[i].select = GUILayout.Toggle(sceneLoadItems[i].select, sceneLoadItems[i].scenePath);
                    GUILayout.EndHorizontal();
                }
                GUILayout.EndScrollView();
                MigrationToolsWindow.DrawLine();
            }
            EditorGUILayout.EndFoldoutHeaderGroup();
        }

        private void LoadAllScenes()
        {
            if (sceneLoadItems == null) sceneLoadItems = new List<SceneLoadItem>();
            else sceneLoadItems.Clear();

            string[] allScenes = AssetDatabase.FindAssets("t:Scene");
            allScenes.ToList().ForEach((string guid) =>
            {
                string path = AssetDatabase.GUIDToAssetPath(guid);
                SceneLoadItem item = new SceneLoadItem(path);
                sceneLoadItems.Add(item);
            });
        }

        private int CountSelectedScenes() {
            int c = 0;
            for (int i = 0; i < sceneLoadItems.Count; i++) {
                if (sceneLoadItems[i].select) c++;
            }
            return c;
        }
    }
}

Editor/SceneOpener.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using UnityEditor;
using UnityEditor.SceneManagement;

namespace Springpunk.MigrationTools.Editor
{
    public class SceneOpener : MigrationTool
    {
        public SceneOpener() : base("Scene Opener") {}

        protected override void OnGUI()
        {
            string[] scenePaths = MigrationToolsWindow.Instance.SceneLoader.SelectedScenePaths;

            GUILayout.Label("Scene opening:");
            if (GUILayout.Button("Open all loaded scenes")) OpenScenes(scenePaths);
            GUILayout.Space(5);
            GUILayout.BeginHorizontal();
            if (GUILayout.Button("Mark all scenes as dirty")) EditorSceneManager.MarkAllScenesDirty();
            if (GUILayout.Button("Save all open scenes")) EditorSceneManager.SaveOpenScenes();
            GUILayout.EndHorizontal();
            GUILayout.Space(5);
            if (GUILayout.Button("Mark scenes as dirty and Save"))
            {
                for (int i = 0; i < EditorSceneManager.sceneCount; i++) EditorSceneManager.MarkAllScenesDirty();
                EditorSceneManager.SaveOpenScenes();
            }
        }

        private static void OpenScenes(string[] scenesToOpen)
        {
            EditorUtility.DisplayProgressBar("Opening scenes...", "Please wait...", 0f);
            int counter = 0;
            foreach (string path in scenesToOpen)
            {
                counter++;
                EditorUtility.DisplayProgressBar("Opening scenes...", path, ((float)counter / scenesToOpen.Length));
                EditorSceneManager.OpenScene(path, OpenSceneMode.Additive);
            }
            EditorUtility.ClearProgressBar();
        }
    }
}

Editor/MigrationTools.cs
using System;
using System.Linq;
using System.Collections.Generic;
using UnityEngine;
using UnityEditor;
using UnityEditor.SceneManagement;
//using TMPro;

namespace Springpunk.MigrationTools.Editor
{
    public class MigrationTools : EditorWindow
    {
        public enum MigrationToolMode
        {
            None = 0,
            CreateSubfolder = 1,
            OpenAndSaveScenes = 2,
            CustomTools = 99,
        }

        private static Action customTools = null;
        public static Action CustomTools
        {
            set => customTools = value;
        }

        private List<string> scenePaths = new List<string>();

        private MigrationToolMode mode = MigrationToolMode.None;
        private bool showScenes = false;
        private string subfolderToCreate = "New Folder";

        private void Awake()
        {
            scenePaths = new List<string>();
        }

        [MenuItem("Window/Springpunk/Migration Tools")]
        public static void ShowWindow()
        {
            MigrationTools mt = GetWindow<MigrationTools>();
            mt.titleContent = new GUIContent("Migration Tools");
        }


        private void OnGUI()
        {
            GUIStyle headStyle = new();
            headStyle.fontSize = 30;
            headStyle.normal.textColor = new Color(0.9f, 0.9f, 0.9f);
            headStyle.margin = new RectOffset(5, 5, 5, 10);

            GUILayout.Label("Migration Tools", headStyle);
            DrawLine();
            GUILayout.Label("Scene loading:");
            DrawSceneSelection();
            DrawLine();
            mode = (MigrationToolMode)EditorGUILayout.EnumPopup("Tool:", mode);
            GUILayout.Space(10);

            switch (mode)
            {
                case MigrationToolMode.CreateSubfolder:
                    DrawFolderCreationTool();
                    break;
                case MigrationToolMode.OpenAndSaveScenes:
                    DrawOpenAllScenesTool();
                    break;
                case MigrationToolMode.CustomTools:
                    DrawCustomTools();
                    break;
                case MigrationToolMode.None:
                default:
                    GUILayout.Label("No tools selected...");
                    break;
            }

            DrawLine();

            //if (GUILayout.Button("Edit TMPs")) {
            //    TMP_Text[] ts = GameObject.FindObjectsOfType<TMP_Text>();
            //    Debug.Log(ts.Length);

            //    foreach (TMP_Text t in ts) t.SetText(t.text + " EDITED");
            //    for (int i = 0; i < EditorSceneManager.sceneCount; i++) {
            //        EditorSceneManager.MarkAllScenesDirty();
            //    }
            //}
        }


        private void DrawLine(float height = 1, float spacing = 10f)
        {
            GUILayout.Space(spacing);
            Rect rect = EditorGUILayout.GetControlRect(false, height);
            rect.height = height;
            EditorGUI.DrawRect(rect, new Color(0.5f, 0.5f, 0.5f));
            GUILayout.Space(spacing);
        }

        private void DrawSceneSelection()
        {
            if (GUILayout.Button("Get All Scenes"))
            {
                scenePaths = GetAllScenes();
            }

            if (scenePaths.Count < 1) return;

            GUILayout.BeginHorizontal();
            GUILayout.Label($"Selected scenes: {scenePaths.Count}");
            showScenes = GUILayout.Toggle(showScenes, "Show:");
            GUILayout.EndHorizontal();
            if (showScenes)
            {
                int i = 0;
                foreach (string scene in scenePaths)
                {
                    i++;
                    GUILayout.Label($"{i} - {scene}");
                }
            }

        }

        private void DrawFolderCreationTool()
        {
            GUILayout.BeginHorizontal();
            GUILayout.Label("Subfolder to create's name:");
            subfolderToCreate = GUILayout.TextField(subfolderToCreate);
            GUILayout.EndHorizontal();
            if (GUILayout.Button("Create subfolders"))
            {
                CreateCustomFolderPerScene(scenePaths.ToArray(), subfolderToCreate);
            }
        }

        private void DrawOpenAllScenesTool()
        {
            if (GUILayout.Button("Open all loaded scenes"))
            {
                OpenScenes(scenePaths.ToArray());
            }
            if (GUILayout.Button("Save all open scenes"))
            {
                EditorSceneManager.SaveOpenScenes();
            }
            if (GUILayout.Button("Mark all scenes as dirty"))
            {
                for (int i = 0; i < EditorSceneManager.sceneCount; i++)
                {
                    EditorSceneManager.MarkAllScenesDirty();
                }
            }

        }

        private void DrawCustomTools()
        {
            if (customTools == null)
            {
                GUILayout.Label("No custom tools available.");
                GUILayout.Label("Assign your custom tools using MigrationTools.CustomTools");
                return;

            }
            customTools.Invoke();
        }

        private static List<string> GetAllScenes()
        {
            string[] allScenes = AssetDatabase.FindAssets("t:Scene");
            List<string> scenePaths = new List<string>();
            allScenes.ToList().ForEach((string guid) =>
            {
                string path = AssetDatabase.GUIDToAssetPath(guid);
                scenePaths.Add(path);
            });

            return scenePaths;
        }

        private static void CreateSceneFolders(string[] scenePaths)
        {

            List<(string, string)> sceneFoldersToCreate = new List<(string, string)>();
            foreach (string path in scenePaths)
            {

                // Get Scene Directory and Name
                int lastSlashIdx = path.LastIndexOf('/');
                string sceneDirectory = path.Substring(0, lastSlashIdx);
                string sceneName = path.Substring(lastSlashIdx + 1).Replace(".unity", "");

                // Mark scenes to create a folder for
                string sceneFolder = $"{sceneDirectory}/{sceneName}";
                if (!AssetDatabase.IsValidFolder(sceneFolder))
                {
                    (string, string) t = (sceneDirectory, sceneName);
                    sceneFoldersToCreate.Add(t);
                }
            }
            if (sceneFoldersToCreate.Count < 1) return;

            Debug.Log($"Creating {sceneFoldersToCreate.Count} missing scene folders.");
            foreach ((string dir, string name) in sceneFoldersToCreate)
                AssetDatabase.CreateFolder(dir, name);
        }

        private static void CreateCustomFolderPerScene(string[] scenePaths, string customFolderName)
        {
            CreateSceneFolders(scenePaths);

            int suffixLength = ".unity".Length;

            List<string> scenesWithMissingSubfolder = new List<string>();
            foreach (string path in scenePaths)
            {
                // Mark scenes to create a folder for
                string sceneFolder = path.Substring(0, path.Length - suffixLength);
                string customFolder = sceneFolder + "/" + customFolderName;
                if (!AssetDatabase.IsValidFolder(customFolder))
                {
                    scenesWithMissingSubfolder.Add(sceneFolder);
                }
            }
            if (scenesWithMissingSubfolder.Count < 1) return;

            Debug.Log($"Creating {customFolderName} subfolder for {scenesWithMissingSubfolder.Count} scenes.");
            foreach (string sceneFolder in scenesWithMissingSubfolder)
                AssetDatabase.CreateFolder(sceneFolder, customFolderName);
        }



        private static void OpenScenes(string[] scenesToOpen)
        {
            EditorUtility.DisplayProgressBar("Opening scenes...", "Please wait...", 0f);
            int counter = 0;
            foreach (string path in scenesToOpen)
            {
                counter++;
                EditorUtility.DisplayProgressBar("Opening scenes...", path, ((float)counter / scenesToOpen.Length));
                EditorSceneManager.OpenScene(path, OpenSceneMode.Additive);
            }
            EditorUtility.ClearProgressBar();
        }
    }
}

Editor/SceneSubfolderCreator.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using UnityEditor;

namespace Springpunk.MigrationTools.Editor
{
    public class SceneSubfolderCreator : MigrationTool
    {
        private string subfolderToCreate = "New Folder";

        public SceneSubfolderCreator() : base("Scene Subfolder Creator") { }

        protected override void OnGUI()
        {
            string[] scenePaths = MigrationToolsWindow.Instance.SceneLoader.SelectedScenePaths;

            GUILayout.BeginHorizontal();
            GUILayout.Label("Subfolder to create's name:");
            subfolderToCreate = GUILayout.TextField(subfolderToCreate);
            GUILayout.EndHorizontal();
            if (GUILayout.Button("Create subfolders"))
            {
                CreateCustomFolderPerScene(scenePaths, subfolderToCreate);
            }
        }

        private static void CreateSceneFolders(string[] scenePaths)
        {

            List<(string, string)> sceneFoldersToCreate = new List<(string, string)>();
            foreach (string path in scenePaths)
            {

                // Get Scene Directory and Name
                int lastSlashIdx = path.LastIndexOf('/');
                string sceneDirectory = path.Substring(0, lastSlashIdx);
                string sceneName = path.Substring(lastSlashIdx + 1).Replace(".unity", "");

                // Mark scenes to create a folder for
                string sceneFolder = $"{sceneDirectory}/{sceneName}";
                if (!AssetDatabase.IsValidFolder(sceneFolder))
                {
                    (string, string) t = (sceneDirectory, sceneName);
                    sceneFoldersToCreate.Add(t);
                }
            }
            if (sceneFoldersToCreate.Count < 1) return;

            Debug.Log($"Creating {sceneFoldersToCreate.Count} missing scene folders.");
            foreach ((string dir, string name) in sceneFoldersToCreate)
                AssetDatabase.CreateFolder(dir, name);
        }

        private static void CreateCustomFolderPerScene(string[] scenePaths, string customFolderName)
        {
            CreateSceneFolders(scenePaths);

            int suffixLength = ".unity".Length;

            List<string> scenesWithMissingSubfolder = new List<string>();
            foreach (string path in scenePaths)
            {
                // Mark scenes to create a folder for
                string sceneFolder = path.Substring(0, path.Length - suffixLength);
                string customFolder = sceneFolder + "/" + customFolderName;
                if (!AssetDatabase.IsValidFolder(customFolder))
                {
                    scenesWithMissingSubfolder.Add(sceneFolder);
                }
            }
            if (scenesWithMissingSubfolder.Count < 1) return;

            Debug.Log($"Creating {customFolderName} subfolder for {scenesWithMissingSubfolder.Count} scenes.");
            foreach (string sceneFolder in scenesWithMissingSubfolder)
                AssetDatabase.CreateFolder(sceneFolder, customFolderName);
        }
    }
}

Example/ExampleMigrationTool.cs
using UnityEngine;
using UnityEditor;

using Springpunk.MigrationTools.Editor;

namespace Springpunk.MigrationTools.Example
{
    [InitializeOnLoad]
    public class ExampleMigrationTool : MigrationTool
    {
        static ExampleMigrationTool()
        {
            // Use this to register a custom tool

            //ExampleMigrationTool example = new ExampleMigrationTool();
            //MigrationToolsWindow.RegisterCustomTool(example);
        }

        public ExampleMigrationTool() : base("Example Migration Tool") { }

        protected override void OnGUI()
        {
            int guess = Random.Range(0, 9);
            GUILayout.Label("Guess the button:");
            GUILayout.Space(10);
            GUILayout.BeginVertical();

            for (int i = 0; i < 9; i += 3)
            {
                GUILayout.BeginHorizontal();
                for (int j = 0; j < 3; j++)
                {
                    if (GUILayout.Button("Pick me"))
                    {
                        if (guess == i + j)
                        {
                            Debug.Log("You guessed correctly!");
                        }
                        else
                        {
                            Debug.Log("Better luck next time...");
                        }
                    }
                }
                GUILayout.EndHorizontal();
            }
            GUILayout.EndVertical();
        }
    }
}

Editor/MigrationToolsWindow.cs
using System;
using System.Linq;
using System.Collections.Generic;
using UnityEngine;
using UnityEditor;
using UnityEditor.SceneManagement;
//using TMPro;

namespace Springpunk.MigrationTools.Editor
{
    public class MigrationToolsWindow : EditorWindow
    {
        public enum MigrationToolMode
        {
            None = 0,
            CreateSubfolder = 1,
            OpenAndSaveScenes = 2,
            CustomTools = 99,
        }

        private static MigrationToolsWindow instance = null;
        public static MigrationToolsWindow Instance {
            get {
                if (instance == null) ShowWindow();
                return instance;
            }
            private set => instance = value;
        }

        private static List<MigrationTool> customTools = null;
        private static List<MigrationTool> CustomTools {
            get {
                if (customTools == null)
                    customTools = new List<MigrationTool>();
                return customTools;
            }
        }

        private MigrationToolMode mode = MigrationToolMode.None;
        private int selectedCustomTool = 0;

        // Tools
        private SceneLoader sceneLoader = new SceneLoader();
        public SceneLoader SceneLoader => sceneLoader;
        private SceneSubfolderCreator subfolderCreator = new SceneSubfolderCreator();
        private SceneOpener sceneOpener = new SceneOpener();


        [MenuItem("Window/Springpunk/Migration Tools")]
        public static void ShowWindow()
        {
            MigrationToolsWindow window = GetWindow<MigrationToolsWindow>();
            window.titleContent = new GUIContent("Springpunk Migration Tools");
            Instance = window;
        }


        private void OnGUI()
        {
            GUIStyle headStyle = new();
            headStyle.fontSize = 30;
            headStyle.normal.textColor = new Color(0.9f, 0.9f, 0.9f);
            headStyle.margin = new RectOffset(5, 5, 5, 10);


            GUILayout.Label("Migration Tools", new GUIStyle {
                fontSize = 30,
                normal = new GUIStyleState
                {
                    textColor = new Color(0.9f, 0.9f, 0.9f),
                },
                margin = new RectOffset(5, 5, 5, 10),
        }
            );

            DrawLine();
            sceneLoader.Draw();
            DrawLine();

            mode = (MigrationToolMode)EditorGUILayout.EnumPopup("Mode:", mode);
            DrawLine();

            switch (mode)
            {
                case MigrationToolMode.CreateSubfolder:
                    subfolderCreator.Draw();
                    break;
                case MigrationToolMode.OpenAndSaveScenes:
                    sceneOpener.Draw();
                    break;
                case MigrationToolMode.CustomTools:
                    DrawCustomTools();
                    break;
                case MigrationToolMode.None:
                default:
                    GUILayout.Label("No tools selected...");
                    break;
            }
        }


        public static void DrawLine(float height = 1, float spacing = 10f)
        {
            GUILayout.Space(spacing);
            Rect rect = EditorGUILayout.GetControlRect(false, height);
            rect.height = height;
            EditorGUI.DrawRect(rect, new Color(0.5f, 0.5f, 0.5f));
            GUILayout.Space(spacing);
        }

        private void DrawCustomTools()
        {
            if (CustomTools.Count < 1)
            {
                GUILayout.Label("No custom tools available.");
                GUILayout.Label($"Use MigrationToolsWindow.RegisterCustomTool(MigrationTool) to register one...");
                return;
            }

            string[] options = CustomTools.ConvertAll((tool) => tool.Name).ToArray();
            selectedCustomTool = EditorGUILayout.Popup("Custom tool:", selectedCustomTool, options);
            DrawLine();

            CustomTools[selectedCustomTool].Draw();
        }

        public static void RegisterCustomTool(MigrationTool tool) {
            if (tool == null) return;
            CustomTools.Add(tool);
        }

        //public static void RegisterCustomTool<T>() where T : MigrationTool
        //{
        //    T tool = new();
        //    if (tool == null) return;
        //    CustomTools.Add(tool);
        //}
    }
}